File: apps/web-services/src/protectedApi_v8/user/topic.ts
/*               "Copyright 2020 Infosys Ltd.
               Use of this source code is governed by GPL v3 license that can be found in the LICENSE file or at https://opensource.org/licenses/GPL-3.0
               This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License version 3" */
import axios from 'axios'
import { Request, Response, Router } from 'express'
import { IGenericApiResponse } from '../../models/generic.model'
import {
  ITopic,
  ITopicResponse,
  ITopicsApiResponse
} from '../../models/topic.model'
import { CONSTANTS } from '../../utils/env'

const apiEndPoints = {
  autocomplete: `${CONSTANTS.ES_BASE}/lex_topic/_search`,
  recommend: `${CONSTANTS.SB_EXT_API_BASE}/v1/topics/recommended?q=new`,
}
export const topicApi = Router()

topicApi.get('/recommend', async (_req: Request, res: Response) => {
  try {
    const topicResponse: ITopicsApiResponse = await axios
      .get<IGenericApiResponse<ITopicsApiResponse>>(apiEndPoints.recommend)
      .then((response) => response.data.result.response)

    const topics: ITopic[] = topicResponse.topics.map(
      (topicsResponse: ITopicResponse) => ({
        count: topicsResponse.count,
        id: topicsResponse.id,
        name: topicsResponse['concepts.name'],
      })
    )
    res.send(topics)
  } catch (err) {
    res.status((err && err.response && err.response.status) || 500).send(err)
  }
})

topicApi.get('/autocomplete', async (req: Request, res: Response) => {
  try {
    const body = {
      _source: { includes: 'name' },
      suggest: {
        'name-suggest': {
          completion: { field: 'name' },
          prefix: req.query.q,
        },
      },
    }

    const response = await axios.request({
      auth: {
        password: CONSTANTS.ES_PASSWORD,
        username: CONSTANTS.ES_USERNAME,
      },
      data: body,
      method: 'POST',
      url: apiEndPoints.autocomplete,
    })
    res.json(response.data)
  } catch (err) {
    res.status((err && err.response && err.response.status) || 500).send(err)
  }
})

File: apps/sb-ext-actors/src/main/java/com/infosys/model/Topic.java
/*               "Copyright 2020 Infosys Ltd.
               Use of this source code is governed by GPL v3 license that can be found in the LICENSE file or at https://opensource.org/licenses/GPL-3.0
               This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License version 3" */
package com.infosys.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.Map;

@JsonIgnoreProperties(ignoreUnknown = true)
public class Topic {
    private String id;
    private String date_created;
    private String date_modified;
    private String name;
    private String source;
    private int source_id;
    private String source_source;
    private String source_status;
    private String status;

    public Topic() {

    }

    public Topic(String id, String date_created, String date_modified, String name, String source,
                 int source_id, String source_source, String source_status, String status) {
        this.id = id;
        this.date_created = date_created;
        this.date_modified = date_modified;
        this.name = name;
        this.source = source;
        this.source_id = source_id;
        this.source_source = source_source;
        this.source_status = source_status;
        this.status = status;
    }

    public static Topic fromMap(Map<String, Object> map) {
        return new ObjectMapper().convertValue(map, Topic.class);
    }

    public Map<String, Object> toMap() {
        return new ObjectMapper().convertValue(this, Map.class);
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getDate_created() {
        return date_created;
    }

    public void setDate_created(String date_created) {
        this.date_created = date_created;
    }

    public String getDate_modified() {
        return date_modified;
    }

    public void setDate_modified(String date_modified) {
        this.date_modified = date_modified;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getSource() {
        return source;
    }

    public void setSource(String source) {
        this.source = source;
    }

    public int getSource_id() {
        return source_id;
    }

    public void setSource_id(int source_id) {
        this.source_id = source_id;
    }

    public String getSource_source() {
        return source_source;
    }

    public void setSource_source(String source_source) {
        this.source_source = source_source;
    }

    public String getSource_status() {
        return source_status;
    }

    public void setSource_status(String source_status) {
        this.source_status = source_status;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }
}
